### PR TITLE
Take minimum size reused buffer

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122643
* #121999
* #124031
* #122825
* #123229
* #122642
* #124030

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -320,6 +320,85 @@ def __init__(
...
+            tensors_to_use.append(mod_tensor)
+            seen_dp.add(dp)
+
+        sorted_tensors = sorted(
+            tensors_to_use, key=lambda x: x.untyped_storage().nbytes()
+        )
+
+        sorted_by_device = defaultdict(list)
+        for t in sorted_tensors:
+            sorted_by_device[t.device].append(t)
```
Comment (peterbell10) [inline torch/_inductor/graph.py:399]: NIT: the `tensors_to_use` list isn't really needed, you can do everything inside the final dict.
```suggestion
            tensors_by_device[mod_tensor.device].append(mod_tensor)
            seen_dp.add(dp)

        for device_tensors in tensors_by_device.values():
            device_tensors.sort(key=lambda x: x.untyped_storage().nbytes())
```

File: torch/_inductor/graph.py [hunk 2]
```diff
@@ -320,6 +320,85 @@ def __init__(
...
+                if type(mod_tensor) is torch.nn.Parameter
+                else mod_tensor
+            )
+            if type(mod_tensor) is not torch.Tensor or not torch._C._has_storage(
+                mod_tensor
+            ):
+                continue
+
+            # MM benchmarking can be sensitive to mantissa
+            # Dont take from integer tensors bc they may be sparse
```
Comment (peterbell10) [inline torch/_inductor/graph.py:382]: sparse -> denormal?

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -320,6 +320,85 @@ def __init__(
...
+
+                if sorted_tensors[mid].untyped_storage().nbytes() > needed_bytes:
+                    min_index = mid
+                    right = mid - 1
+                else:
+                    left = mid + 1
+
+            return min_index
+
+        min_index = binary_search_needed_size()
```
Comment (peterbell10) [inline torch/_inductor/graph.py:348]:
```suggestion
        min_index = bisect.bisect_left(
            sorted_tensors, needed_bytes, key=lambda x: x.untyped_storage().nbytes()
        )
```

Comment (eellison) [reply torch/_inductor/graph.py:352]: Wow, didn't know about this, nice !

**Review by shunting314 — COMMENTED** [review]
Remind me again why we need reuse module parameter's storage for benchmarking?

We are running benchmarking kernels serially any ways, why not allocating new tensors ?

**Comment by eellison** [comment]
@shunting314 let me test memory numbers with and without these two prs. it may not be needed

File: torch/_inductor/graph.py [hunk 4]
```diff
@@ -335,6 +336,74 @@ def __init__(
...
+        min_index = bisect.bisect_left(
+            sorted_tensors, needed_bytes, key=lambda x: x.untyped_storage().nbytes()
+        )
+        if min_index == len(sorted_tensors):
+            return None
+
+        while (
+            min_index < len(sorted_tensors)
+            and sorted_tensors[min_index].untyped_storage().data_ptr()
+            in self.module_storages_in_benchmark_use
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/graph.py:364]: we add a weakref finalizer to the tensor we construct from the data pointer

**Pull request closed without merging**